src/app/components/table-of-content/table-of-content.component.ts

```diff
@@ -49,4 +115,33 @@ export class TableOfContentComponent {
   dataSource = new MatTreeFlatDataSource(this.treeControl, this.treeFlattener);
 
   hasChild = (_: number, node: HeaderTreeNode) => node.expandable;
+
+  private createOffsets(data: HeaderNode[], array = this.offsetHeaders): void {
+    for (const element of data) {
+      array.push(element.id);
+
+      if (element.children.length) {
+        this.createOffsets(element.children, array);
+      }
+    }
+  }
```

Comment:
I would rather not mutate an argument, so you can make this method to return the offsetHeaders instead:
```suggestion
  private createOffsets(data: HeaderNode[]): string[] {
    return data.reduce((offsetHeaders: string[], element: HeaderNode) => [
      ...offsetHeaders,
      element.id,
      ...(element.children.length ? this.createOffsets(element.children) : []),
    ], []);
  }
```